### ISSUE TITLE
NVDA doesn't know a difference between Links and Same Page Links

### ISSUE THREAD

**Issue by nvdaTrac** [nvaccess/test_nvdaTracImport#221]
**Reported by hkatic on 23 Jul 2008 09:11 UTC**
**Blocking #3434, #5190**
In virtual buffers, NVDA doesn't know a difference between external and same page links. So for example, any user including myself may be confuzed whyle viewing a Web page in FireFox for example where lots of links exist, because he or she may not know is this link going to another page or to any other place on the same page. So, it is required to make NVDA speaking "same page link" before any link which moves to a different part of the same page e.g. "same page link system requirements". The same applyes to FTP links and Send Mail links.

**Comment by nvdaTrac** [comment]
**Comment by jteh on 23 Jul 2008 23:32 UTC**
This will not be done for 0.6p2.

Currently, announcement of same page links is quite difficult to implement because Firefox does not provide an easy way for us to do this.

Also, consider the fact that most users who do not use a screen reader actually can't differentiate between different types of links without looking at the URL, unless the site marks them in some way. If a site uses a graphic to denote external links, this should be obvious to a screen reader user also. As far as I know, same page links, mailto links, etc. do not look any different to any other link unless they are styled differently.

**Comment by nvdaTrac** [comment]
**Comment by norrumar on 12 Feb 2011 12:44 UTC**
I have used JAWS before NVDA, and JAWS announces different types of links: links on the same page, transfer files (ftp://), etc. I don't know how NVDA source code is built. But if the value of href attribute, in <a> element, could be used, NVDA may announce, for instance, "Link on the same page" when href value begins: "#". I think that this enhancement would be useful.

**Comment by nvdaTrac** [comment]
**Comment by jteh on 12 Feb 2011 23:05 UTC**
As noted in comment:1, I don't see why a screen reader should inform the user of this. Links don't look any different unless the site marks them differently (in which case this should be indicated in an alternative way for accessibility); they are just links. If this does get implemented, I certainly think it should be configurable and disabled by default.

**Comment by nvdaTrac** [comment]
**Comment by jteh on 12 Feb 2011 23:14 UTC**
@norrumar: Did you actually mean to accept this ticket or was that an error?

**Comment by nvdaTrac** [comment]
**Comment by norrumar on 15 Feb 2011 07:46 UTC**
I don¡t know the meaning of accepting a ticket, excuse me. About the option for wnowing the types of links, I think that implementing it as a configurable option is a good idea. I thing also that a screen reader, sometimes, sould implement task or options, although they ar not provided using accesibility or visual properties, for example ARIA or css. Links that points to the same page can be often created due to accesibility (or usability), because they get an easier navigation. Hotkeys (accesskey) perhaps are not visible on web pages, and so some web sites include an accesibility page, where these keys are represented on a table. Furthermore, if you can see the screen, I think that knowing the type of a link is easier: if it's a link to the same page, you can see the content that is pointed by the link, and the link text can be a reference for knowing its type. Excuse me for my bad English. Here is a link to a WCAG 2.0 document, because I can't explain you better what  want to say:
http://www.w3.org/WAI/WCAG20/quickref/#qr-navigation-mechanisms-skip

IThanks.

**Comment by nvdaTrac** [comment]
**Comment by norrumar on 17 Apr 2011 12:44 UTC**
Replying to [hkatic](ticket:141):
I think that it would be useful that NVDA can read the type of link in virtual buffers without plugins, if possible. I have attached a plugin for this requirement, but You have to press control+shift+a, and when a link is focused, NVDA reads the type. I have developed the plugin in Spanish. If you are reading another element, not a link, and you press control+shift+a, NVDA could announce the type of the previous link, if it is focused although the cursor is not over that link. I have included five possible messages (in Spanish):
- Anchor: If the destination is a part of the same or another document, usually the same.
- Link to the same document.
- Mail link.
- FTP link.
- External link: For links to external resources.
Furthermore, this script can:
- Announce the link URI, for instance: http://www.example.com#somewhere: Pressing control+shift+a two times.
- Copy the URI to clipboard: Pressing control+shift+a three times.
I have tested this plugin on Firefox, Internet Explorer 8 and Chrome. The script doesn't work on Chrome.

> In virtual buffers, NVDA doesn't know a difference between external and same page links. So for example, any user including myself may be confuzed whyle viewing a Web page in FireFox for example where lots of links exist, because he or she may not know is this link going to another page or to any other place on the same page. So, it is required to make NVDA speaking "same page link" before any link which moves to a different part of the same page e.g. "same page link system requirements". The same applyes to FTP links and Send Mail links.

**Comment by nvdaTrac** [comment]
**Comment by jteh on 13 Feb 2013 07:47 UTC**
Replying to [norrumar](comment:6):
> I thing also that a screen reader, sometimes, sould implement task or options, although they ar not provided using accesibility or visual properties
I don't see why a screen reader should read anything that isn't available to any other user, sighted or otherwise.

> Links that points to the same page can be often created due to accesibility (or usability), because they get an easier navigation.
True, but these always say "Skip to content", "Return to top" or whatever, so what they do is obvious from the text. Saying "same page link" is just unnecessary verbosity.

> Hotkeys (accesskey) perhaps are not visible on web pages, and so some web sites include an accesibility page, where these keys are represented on a table.
I don't quite follow how this is related. If it's a separate page, it's not a same page link.

> Furthermore, if you can see the screen, I think that knowing the type of a link is easier: if it's a link to the same page, you can see the content that is pointed by the link
You can't if the content isn't on the screen (i.e. you have to scroll), which is the primary use of same page links.
> and the link text can be a reference for knowing its type.
That's also true for screen reader users: the text indicates what the link will do.

**Comment by nvdaTrac** [comment]
**Comment by mohammed on 30 Aug 2014 13:56 UTC**
in page links (anchors) are almost always styled differently. Sighted developers or followers  can correct me. 

it's one part of NVDA that I couldn't get used to yet. I think differentiating between normal and in page links can make our lives easier. so consider this just a vote for implimenting this. if you're worried for verbosity we can probably wait until NVDA implements audio themes.
